### PR TITLE
ENH: add optional ddof param to DataFrame.cov()

### PR DESCRIPTION
**Pull request by edhalter**
- [ ] closes #xxxx (searched, didn't see any issues on this topic)
 - [x] tests added / passed
 - [x] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff``
 - [x] whatsnew entry

Noticed that the .cov() methods on DataFrame and Series assume the user wants to calculate sample variance and normalize by N-1. In contrast, the .var() and .std() methods on DataFrame, Series, DataFrameGroupBy, and SeriesGroupBy take a ddof parameter. Its default value is 1, calculating the sample std/var, but the user can specify ddof=0 to calculate the population std/var.

Simply added a ddof param (default value 1) to the .cov() methods. The value is then passed into other method calls as required. Did have to modify 1 bit of Cython code in algos.pyx, the nancorr() function.

### CODE REVIEW

File: pandas/core/frame.py [hunk 1]
```diff
@@ -4968,15 +4968,20 @@ def corr(self, method='pearson', min_periods=1):
...
+        Normalized by N-1 by default. This can be changed using the ddof argument"
 
         Parameters
         ----------
         min_periods : int, optional
             Minimum number of observations required per pair of columns
             to have a valid result.
+        ddof : int, default 1
+            Delta Degrees of Freedom.  The divisor used in calculations
+            is ``N - ddof``, where ``N`` represents the number of elements.
```
Comment (jreback) [inline pandas/core/frame.py]: versionadded tag

File: pandas/core/series.py [hunk 2]
```diff
@@ -1374,15 +1374,20 @@ def corr(self, other, method='pearson', min_periods=None):
...
+        Normalized by N-1 by default. This can be changed using the ddof argument"
 
         Parameters
         ----------
         other : Series
         min_periods : int, optional
             Minimum number of observations needed to have a valid result
+        ddof : int, default 1
+            Delta Degrees of Freedom.  The divisor used in calculations
+            is ``N - ddof``, where ``N`` represents the number of elements.
```
Comment (jreback) [inline pandas/core/series.py]: same

File: doc/source/whatsnew/v0.21.0.txt [hunk 3]
```diff
@@ -27,6 +27,9 @@ New features
 Other Enhancements
 ^^^^^^^^^^^^^^^^^^
 
+- The ``.cov()`` methods for DataFrame and Series now accept the parameter
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: add this PR number here

Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: accept -> accepts

**Review by jreback — APPROVED** [review]
lgtm.

File: pandas/core/frame.py [hunk 4]
```diff
@@ -4968,15 +4968,20 @@ def corr(self, method='pearson', min_periods=1):
 
         return self._constructor(correl, index=idx, columns=cols)
 
-    def cov(self, min_periods=None):
+    def cov(self, min_periods=None, ddof=1):
         """
-        Compute pairwise covariance of columns, excluding NA/null values
+        Compute pairwise covariance of columns, excluding NA/null values.
+        
+        Normalized by N-1 by default. This can be changed using the ddof argument"
```
Comment (TomAugspurger) [inline pandas/core/frame.py]: I don't think this line is needed (as it's have it in the Notes section)

File: pandas/core/frame.py [hunk 5]
```diff
@@ -4985,7 +4990,8 @@ def cov(self, min_periods=None):
         Notes
         -----
         `y` contains the covariance matrix of the DataFrame's time series.
-        The covariance is normalized by N-1 (unbiased estimator).
+        The covariance is normalized by N-ddof -- for the default ddof value of
```
Comment (TomAugspurger) [inline pandas/core/frame.py]: Double backticks around `N-ddof` and `ddof`.

File: pandas/core/series.py [hunk 6]
```diff
@@ -1374,15 +1374,20 @@ def corr(self, other, method='pearson', min_periods=None):
         return nanops.nancorr(this.values, other.values, method=method,
                               min_periods=min_periods)
 
-    def cov(self, other, min_periods=None):
+    def cov(self, other, min_periods=None, ddof=1):
         """
         Compute covariance with Series, excluding missing values
+        
+        Normalized by N-1 by default. This can be changed using the ddof argument"
```
Comment (TomAugspurger) [inline pandas/core/series.py]: Can probably make this a new Notes section, and combine with the note that's currently in the Returns section.

**Comment by jreback** [comment]
can you rebase and update?

**Comment by edhalter** [comment]
Made changes to docstrings, etc. based on comments. Rebased and updated.

**Comment by TomAugspurger** [comment]
@edhalter did you push your changes? There seems to be a conflict in the whatsnew file. You may need to force push to your branch.

**Comment by edhalter** [comment]
Already pushed -- the conflicting file has my most recent changes. Someone just needs to edit it and append my new line to the other new lines under "Other Enhancements".

**Comment by jreback** [comment]
@edhalter pls rebase on master. this will present a conflict which you fix, then push.

**Comment by edhalter** [comment]
Rebased on master.

File: pandas/core/frame.py [hunk 7]
```diff
@@ -5005,15 +5005,20 @@ def corr(self, method='pearson', min_periods=1):
...
-        Compute pairwise covariance of columns, excluding NA/null values
+        Compute pairwise covariance of columns, excluding NA/null values.
 
         Parameters
         ----------
         min_periods : int, optional
             Minimum number of observations required per pair of columns
             to have a valid result.
+        ddof : int, default 1
+            .. versionadded:: 0.21.0
```
Comment (jreback) [inline pandas/core/frame.py]: move the versionadded tag below the expl.

File: pandas/core/frame.py [hunk 8]
```diff
@@ -5022,7 +5027,8 @@ def cov(self, min_periods=None):
         Notes
         -----
         `y` contains the covariance matrix of the DataFrame's time series.
-        The covariance is normalized by N-1 (unbiased estimator).
+        The covariance is normalized by ``N-ddof`` -- for the default ``ddof``
```
Comment (jreback) [inline pandas/core/frame.py]: just start a new sentence, I don't think '--' works here.

File: pandas/core/series.py [hunk 9]
```diff
@@ -1408,18 +1408,26 @@ def cov(self, other, min_periods=None):
         other : Series
         min_periods : int, optional
             Minimum number of observations needed to have a valid result
+        ddof : int, default 1
+            .. versionadded:: 0.21.0
```
Comment (jreback) [inline pandas/core/series.py]: same as above

File: pandas/core/series.py [hunk 10]
```diff
@@ -1408,18 +1408,26 @@ def cov(self, other, min_periods=None):
...
+            is ``N - ddof``, where ``N`` represents the number of elements.
 
         Returns
         -------
         covariance : float
 
-        Normalized by N-1 (unbiased estimator).
+        Notes
+        -----
+        The returned covariance is normalized by ``N-ddof`` -- for the default
```
Comment (jreback) [inline pandas/core/series.py]: same as above

File: pandas/_libs/algos.pyx [hunk 11]
```diff
@@ -325,7 +330,7 @@ def nancorr(ndarray[float64_t, ndim=2] mat, bint cov=0, minp=None):
                             sumxx += vx * vx
                             sumyy += vy * vy
 
-                    divisor = (nobs - 1.0) if cov else sqrt(sumxx * sumyy)
```
Comment (jreback) [inline pandas/_libs/algos.pyx]: what about ddofv >= nobs case? can you add a test for this. looks like divisor == 0 is handled, but what about ddofv > nobs?

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by edhalter** [comment]
Added some checks in DataFrame.cov() and the nancov() func called by Series.cov():

- sanity check on input ddof
- warn when N - ddof <= 0

For the latter, alternatively, could throw an Exception, or just silently return nan. The warning seemed better -- that's what numpy's cov() does:

`warnings.warn("Degrees of freedom <= 0 for slice", RuntimeWarning)`

However, the Cython nancorr() func called by DataFrame.cov() does just silently return nan.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
> warnings.warn("Degrees of freedom <= 0 for slice", RuntimeWarning)

can you change UserWarning -> RuntimeWarning to be consistent

> However, the Cython nancorr() func called by DataFrame.cov() does just silently return nan.

can you change (and test) this as well?

**Comment by jreback** [comment]
if you can rebase and update

**Comment by jreback** [comment]
can you rebase

**Comment by edhalter** [comment]
Returning to this.... Changed warning type to RuntimeWarning, and added a single warning that will throw in cython from nancorr() if at least 1 result in the matrix is set to NaN due to N - ddof <= 0.

Manual tests looked good. Also added some more automated tests in DataFrame's test_cov().

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/frame.py [hunk 12]
```diff
@@ -5307,23 +5312,33 @@ def cov(self, min_periods=None):
...
+
         numeric_df = self._get_numeric_data()
         cols = numeric_df.columns
         idx = cols.copy()
         mat = numeric_df.values
 
         if notna(mat).all():
+            fill_nan = False
             if min_periods is not None and min_periods > len(mat):
+                fill_nan = True
```
[the remark this answers is not in the document]
Comment (jreback) [reply pandas/core/frame.py]: can you explain whats going on here

File: pandas/tests/frame/test_analytics.py [hunk 13]
```diff
@@ -129,31 +129,55 @@ def test_corr_cov_independent_index_column(self):
...
+        df_na = self.frame.copy()
+        num_rows = df_na.shape[0]
+        num_isna = num_rows // 2
+        num_notna = num_rows - num_isna
+        df_na['A'][:num_isna] = nan
+        df_na['B'][num_isna:2 * num_isna] = nan
+
+        # compare each number in DataFrame.cov() matrix
+        # to corresponding Series.cov() number
+        for ddof in [0, 1]:
```
[the remark this answers is not in the document]
Comment (jreback) [reply pandas/tests/frame/test_analytics.py]: can you do this

**Review by jreback — CHANGES_REQUESTED** [review]
pls rebase and update according to comments

**Comment by jreback** [comment]
closing as stale, but if you'd like to continue work, pls rebase and ping.

**Pull request closed without merging**